src/Operations.ts

```diff
@@ -309,6 +309,74 @@ export class Operations {
     }
   }
 
+  public setRowOrder(sheetId: number, rowMapping: [number, number][]): void {
+    const buffer: [SimpleCellAddress, ClipboardCell][][] = []
+    for(const [source, target] of rowMapping ) {
+      if(source!==target) {
+        const rowRange = AbsoluteCellRange.spanFrom({sheet: sheetId, col: 0, row: source}, Infinity, 1)
+        if (this.dependencyGraph.matrixMapping.isFormulaMatrixInRange(rowRange)) {
+          throw new SourceLocationHasMatrixError()
+        }
```

Comment:
I think there it can be done like with other cruds with `ensureItIsPossible...` method
1. `testRowOrderForMatrices` can be changed in such way it will throw error
2. `testRowOrderForMatrices` can be moved to CrudOperations and called beffore `operations.setRowOrder`
3. `HyperFormula.isItPossibleToSetRowOrder` will catch error anyway and return bool

Reply:
Moreover:
1. Less unnecessary checks during undo redo
2  Matrix existence will be checked in separate loop before breaking any numeric matrices, so you won't loose numeric matrix optimizations in case of error in the middle of iteration